optimum/commands/optimum_cli.py

```diff
@@ -105,34 +64,57 @@ def resolve_command_to_command_instance(
     return command2command_instance
 
 
-def dynamic_load_commands_in_register() -> (
+def load_optimum_namespace_cli_commands() -> (
     List[Tuple[Union[Type[BaseOptimumCLICommand], CommandInfo], Optional[Type[BaseOptimumCLICommand]]]]
 ):
     """
-    Loads a list of command classes to register to the CLI from the `optimum/commands/register/` directory.
-    It will look in any python file if there is a `REGISTER_COMMANDS` list, and load the commands to register
-    accordingly.
-    At this point, nothing is actually registered in the root CLI.
+    Loads a list of command classes to register to the CLI from the `optimum.commands.register` namespace of each optimum subpackage/distribution.
 
     Returns:
         `List[Tuple[Union[Type[BaseOptimumCLICommand], CommandInfo], Optional[Type[BaseOptimumCLICommand]]]]`: A list
         of tuples with two elements. The first element corresponds to the command to register, it can either be a
         subclass of `BaseOptimumCLICommand` or a `CommandInfo`. The second element corresponds to the parent command,
         where `None` means that the parent command is the root CLI command.
     """
-    commands_to_register = []
-    register_dir_path = Path(__file__).parent / "register"
-    for filename in register_dir_path.iterdir():
-        if filename.is_dir() or filename.suffix != ".py":
-            if filename.name not in ["__pycache__", "README.md"]:
-                logger.warning(
-                    f"Skipping {filename} because only python files are allowed when registering commands dynamically."
-                )
+    command_registration_files = []
+    for dist in importlib.metadata.distributions():
+        if dist.metadata["Name"] is None:
+            continue
+        if dist.metadata["Name"] == "optimum-benchmark":
+            continue
+        if not dist.metadata["Name"].startswith("optimum"):
+            # it might be better (and more secure ?) to use an explicit list of optimum subpackages here
+            continue
+
+        if dist.metadata["Name"] == "optimum":
+            # NOTE: optimum no longer registers commands in the main package, we do this for testing purposes only
+            # optimum can't have an __init__.py file, so we use optimum.pipelines instead
+            dist_name = "optimum.pipelines"
```

Comment:
why do we need this ?

Reply:
there's a test in tests/cli that puts a register_custom_command.py in optimum/commands/register and for it to work we need to the location of optimum but since the package sees 'optimum' as a namespace it doesn't really have a location (it has multiple locations, one in every distribution) so we use optimum.pipelines.
this hack and the test i'm talking about are not really needed at this point, because now we do test command registration by installing optimum and optimum-onnx. I can remove both the hack and the custom command test wdyt ?